ledger/acctupdates.go

```diff
@@ -376,60 +523,537 @@ func (au *accountUpdates) getAssetCreatorForRound(rnd basics.Round, aidx basics.
+	}
+
+	au.roundTotals = append(au.roundTotals, newTotals)
+}
+
+func (au *accountUpdates) totals(rnd basics.Round) (totals AccountTotals, err error) {
+	au.accountsMu.RLock()
+	defer au.accountsMu.RUnlock()
+	offset, err := au.roundOffset(rnd)
+	if err != nil {
+		return
+	}
+
+	totals = au.roundTotals[offset]
+	return
+}
+
+func (au *accountUpdates) getCatchpointStream(round basics.Round) (io.ReadCloser, error) {
+	dbFileName := ""
+	err := au.dbs.rdb.Atomic(func(tx *sql.Tx) (err error) {
+		dbFileName, _, _, err = getCatchpoint(tx, round)
+		return
+	})
+	if err != nil && err != sql.ErrNoRows {
+		// we had some sql error.
+		return nil, fmt.Errorf("accountUpdates: getCatchpointStream: unable to lookup catchpoint %d: %v", round, err)
+	}
+	if dbFileName != "" {
+		catchpointPath := filepath.Join(au.dbDirectory, dbFileName)
+		file, err := os.OpenFile(catchpointPath, os.O_RDONLY, 0666)
+		if err == nil && file != nil {
+			return file, nil
+		}
+		// else, see if this is a file-not-found error
+		if os.IsNotExist(err) {
+			// the database told us that we have this file.. but we couldn't find it.
+			// delete it from the database.
+			err := au.saveCatchpointFile(round, "", 0, "")
+			if err != nil {
+				au.log.Warnf("accountUpdates: getCatchpointStream: unable to delete missing catchpoint entry: %v", err)
+				return nil, err
+			}
+
+			return nil, ErrNoEntry{}
+		}
+		// it's some other error.
+		return nil, fmt.Errorf("accountUpdates: getCatchpointStream: unable to open catchpoint file '%s' %v", catchpointPath, err)
+	}
+
+	// if the database doesn't know about that round, see if we have that file anyway:
+	fileName := filepath.Join("catchpoints", catchpointRoundToPath(round))
+	catchpointPath := filepath.Join(au.dbDirectory, fileName)
+	file, err := os.OpenFile(catchpointPath, os.O_RDONLY, 0666)
+	if err == nil && file != nil {
+		// great, if found that we should have had this in the database.. add this one now :
+		fileInfo, err := file.Stat()
+		if err != nil {
+			// we couldn't get the stat, so just return with the file.
+			return file, nil
+		}
+
+		err = au.saveCatchpointFile(round, fileName, fileInfo.Size(), "")
+		if err != nil {
+			au.log.Warnf("accountUpdates: getCatchpointStream: unable to save missing catchpoint entry: %v", err)
+		}
+		return file, nil
+	}
+	return nil, ErrNoEntry{}
+}
+
+// functions below this line are all internal functions
+
+// accountHashBuilder calculates the hash key used for the trie by combining the account address and the account data
+func accountHashBuilder(addr basics.Address, accountData basics.AccountData, encodedAccountData []byte) []byte {
+	hash := make([]byte, 4+crypto.DigestSize)
+	// write out the lowest 32 bits of the reward base. This should improve the caching of the trie by allowing
+	// recent updated to be in-cache, and "older" nodes will be left alone.
+	for i, rewards := 3, accountData.RewardsBase; i >= 0; i, rewards = i-1, rewards>>8 {
+		hash[i] = byte(rewards & 0x255)
```

Comment:
This should be `0xFF`

Reply:
actually, you can probably just kill the bitwise and entirely: https://play.golang.org/p/7bgJ5VOYh_C

Reply:
I have no idea how you managed to pick 12345678, where 12345678 & 255 is 78.
i.e. the  12345600 is divisible by 256. what are the chances for that ?!